src/modules-arch/module-descriptor.js

```diff
@@ -0,0 +1,198 @@
+   * anything like require(SomeModule.componentPath). To make the components
+   * lazy-loaded we have to define them in a way like _ => import('./path)
+   * directly during the definition.
+   *
+   * @param {String} [opts.importStoreModule]
+   * Path to the entry point vuex-module of the module. Please provide
+   * function like _ => import('./path).
+   *
+   * @param {ModuleDescriptor.constructor[]} [opts.dependencies]
+   * Constructors of modules the module depends on.
+   *
+   * @param {ModuleDescriptor.constructor[]} [opts.allowedSubmodules]
+   * Constructors of modules the module could use as submodules.
+   *
+   * @param {ModuleDescriptor[]} [opts.submodules]
+   * Modules that the module will use as submodules.
+   * Cannot contain any entries except those described in
+   * opts.allowedSubmodules.
+   *
+   * @param {ModuleDescriptor.constructor[]} [opts.incompatibles]
+   * Constructors of modules that cannot be used simultaneously with the module.
+   *
+   *
+   * @param {Boolean} [opts.isCorporateOnly]
+   * If `true` the component will be accessible only by corporate accounts
+   *
+   */
+  constructor (opts = {}) {
+    // will be assigned automatically while the component renders
+    this._createdComponentUid = -1
+    this._createdComponent = {}
+
+    const {
+      dependencies = [],
+      allowedSubmodules = [],
+      submodules = [],
+      incompatibles = [],
+      importComponent = null,
+      importStoreModule = null,
+    } = opts
+
+    if (typeof importComponent !== 'function') {
+      throw new Error(`${this.constructor.name}: no importComponent provided`)
+    }
+
+    this._importComponent = this._rememberUidAfterImport(importComponent)
+
+    if (importStoreModule) {
+      this._importStoreModule = importStoreModule
+    }
+
+    this._dependencies = dependencies
+    this._incompatibles = incompatibles
+    this._allowedSubmodules = allowedSubmodules
+    this._isCorporateOnly = opts.isCorporateOnly || false
+
+    this.validateSubmodules(submodules)
+    this._submodules = submodules
+  }
+
+  get importComponent () { return this._importComponent }
+  get importStoreModule () { return this._importStoreModule }
+  get dependencies () { return this._dependencies }
+  get allowedSubmodules () { return this._allowedSubmodules }
+  get submodules () { return this._submodules }
+  get incompatibles () { return this._incompatibles }
+  get isCorporateOnly () { return this._isCorporateOnly }
+  get createdComponentUid () { return this._createdComponentUid }
+  get createdComponent () { return this._createdComponent }
+  get isComponentCreated () { return this._createdComponentUid >= 0 }
+
+  _rememberUidAfterImport (importFn) {
+    return async _ => {
+      if (this.isComponentCreated) {
+        return importFn()
+      }
+
+      const imported = await importFn()
+      const componentDefinition = Object.values(imported)
+        .find(item => item.render)
```

Comment:
1. the problem is that we import plain JS object here. I don’t want to do something like `new Vue()` because it executes all the `beforeCreated` hooks. Just gonna add some other conditions to make it a bit less risky.

2. both import() and require() return { default: {...} } if you export them as `export default` to extract `default` we should `await` while `import` or `require` done their job and do it like `result.default`